packages/backend-common/src/reading/BitbucketUrlReader.ts

```diff
@@ -115,74 +79,39 @@ export function getApiUrl(target: string, provider: ProviderConfig): URL {
   }
 }
 
-export function readConfig(config: Config): ProviderConfig[] {
-  const providers: ProviderConfig[] = [];
-
-  const providerConfigs =
-    config.getOptionalConfigArray('integrations.bitbucket') ?? [];
+/**
+ * A processor that adds the ability to read files from Bitbucket v1 and v2 APIs, such as
+ * the one exposed by Bitbucket Cloud itself.
+ */
+export class BitbucketUrlReader implements UrlReader {
+  private readonly config: BitbucketIntegrationConfig;
 
-  // First read all the explicit providers
-  for (const providerConfig of providerConfigs) {
-    const host = providerConfig.getOptionalString('host') ?? 'bitbucket.org';
-    let apiBaseUrl = providerConfig.getOptionalString('apiBaseUrl');
-    const token = providerConfig.getOptionalString('token');
-    const username = providerConfig.getOptionalString('username');
-    const appPassword = providerConfig.getOptionalString('appPassword');
+  static factory: ReaderFactory = ({ config }) => {
+    const configs = readBitbucketIntegrationConfigs(
+      config.getOptionalConfigArray('integrations.bitbucket') ?? [],
+    );
+    return configs.map(provider => {
+      const reader = new BitbucketUrlReader(provider);
+      const predicate = (url: URL) => url.host === provider.host;
+      return { reader, predicate };
+    });
+  };
 
-    if (apiBaseUrl) {
-      apiBaseUrl = apiBaseUrl.replace(/\/+$/, '');
-    } else if (host === 'bitbucket.org') {
-      apiBaseUrl = DEFAULT_BASE_URL;
-    }
+  constructor(config: BitbucketIntegrationConfig) {
+    const { host, apiBaseUrl, token, username, appPassword } = config;
 
     if (!apiBaseUrl) {
       throw new Error(
         `Bitbucket integration for '${host}' must configure an explicit apiBaseUrl`,
       );
     }
```

Comment:
these checks now typically happen in the constructors or methods of the class that does the actual work instead of in the config parsing - since different implementations may be in need of different subsets of the config